### PR TITLE
TSDB backfill: enforce # EOF

### LINKED ISSUE #8228: promtool tsdb backfil does not check for # EOF
If I create a file without a newline, we still ingest the metrics.

```
echo -n a 1 $(date +%s) > test
```

```
./promtool tsdb create-blocks-from openmetrics test data4
BLOCK ULID                  MIN TIME                       MAX TIME                       DURATION  NUM SAMPLES  NUM CHUNKS  NUM SERIES  SIZE
01ER321F9RE4P1QVSC8KD1KJZE  2020-11-26 19:56:14 +0000 UTC  2020-11-26 19:56:14 +0000 UTC  1ms       1            1           1           514B
```

I do think that we should enfore the files to end with \n, to backfilling corruption (e.g a truncated timestamp). I do not especially think we should enforce the # EOF.

### PR DESCRIPTION
**Pull request by roidelapluie**
cc @aSquare14 

Fix #8228

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: cmd/promtool/backfill.go [hunk 1]
```diff
@@ -51,6 +52,16 @@ func (p *OpenMetricsParser) Next() (textparse.Entry, error) {
 		if et, err := p.Parser.Next(); err != io.EOF {
 			return et, err
 		}
+		if string(line) == "# EOF\n" {
```
Comment (brian-brazil) [inline cmd/promtool/backfill.go:55]: The \n is optional.

Comment (roidelapluie) [reply cmd/promtool/backfill.go:55]: We add it here: https://github.com/prometheus/prometheus/pull/8232/files#diff-1f356c3850c952097aff258debcd0f10d6a313b616a17510de7d6c0448b4d853R49

**Review by aSquare14 — APPROVED** [review]

File: cmd/promtool/backfill.go (hunk 1 above)
Comment (brian-brazil) [reply cmd/promtool/backfill.go:55]: That's not obvious, the unittests need to cover this.

In general I think it's unwise to hack the parsing like this - it bypasses not just this check, but future checks that will work across lines.

**Review by beorn7 — APPROVED** [review]
Let's go with this "hacky" solution for now to stop the bleeding (the current implementation doesn't properly honor the EOF lines) and let's approach a more elegant but also more involved solution for the parsing problem separately (either by switching to an `mmap`-based approach or giving the parser itself the capability to parse from a `Reader`).

@roidelapluie perhaps add a TODO along those lines?

**Comment by beorn7** [comment]
#8266 for tracking the parsing issue.

**Comment by brian-brazil** [comment]
I still am not convinced that this is the right approach to be taking here. We'd be adding a workaround on to a workaround to duplicate logic we have elsewhere. I believe that taking advantage of mmap here via our existing tsdb utility library would be simpler and safer, as we would be able to get rid of NewOpenMetricsParser and use the parser code with the []byte it expects directly.

I'm not sure what your technical concerns are in relation to mmap, considering this particular use case. Issues we have to consider in the TSDB don't apply to a CLI tool like this.

This still also lacks the unittests I requested to verify it handles all the cases correctly.

**Comment by beorn7** [comment]
This is not about technical concerns, this is about fixing the bug _now_ and do it nicer later rather than leaving the bug in the wild until we have a better solution in place.

**Comment by roidelapluie** [comment]
I have pushed a mmap version

**Comment by brian-brazil** [comment]
:+1: 

That looks better alright, thanks.

**Comment by beorn7** [comment]
Great. I was mostly afraid we would go into another week-long review cycle with the mmap solution.

**Pull request merged**